File: ios/Sources/FindUsEngine/DynamicGraph.swift
import Foundation

/// Device-local dynamic graph (port of core/graph.py): holds only THIS
/// device's heard neighborhood — O(k), never O(N).
public final class NodeEntry {
    public let id: NodeId
    public var lifecycle: NodeLifecycle
    public var lastSeen: Double
    public let firstSeen: Double
    public init(id: NodeId, lifecycle: NodeLifecycle, lastSeen: Double, firstSeen: Double) {
        self.id = id; self.lifecycle = lifecycle
        self.lastSeen = lastSeen; self.firstSeen = firstSeen
    }
}

public final class EdgeEntry {
    public let endpoints: (NodeId, NodeId)
    public var lifecycle: EdgeLifecycle
    public let firstSeen: Double
    public var lastSeen: Double
    public var lastBidirectionalExchange: Double
    public init(endpoints: (NodeId, NodeId), lifecycle: EdgeLifecycle,
                firstSeen: Double, lastSeen: Double, lastBidirectionalExchange: Double = 0) {
        self.endpoints = endpoints; self.lifecycle = lifecycle
        self.firstSeen = firstSeen; self.lastSeen = lastSeen
        self.lastBidirectionalExchange = lastBidirectionalExchange
    }
}

public final class DynamicGraph {
    public let nodeStaleAfter: Double
    public let nodeExpireAfter: Double
    public let edgeStaleAfter: Double
    public let edgeExpireAfter: Double

    private var nodes: [String: NodeEntry] = [:]
    private var edges: [String: EdgeEntry] = [:]

    public init(nodeStaleAfter: Double = 60.0, nodeExpireAfter: Double = 120.0,
                edgeStaleAfter: Double = 60.0, edgeExpireAfter: Double = 60.0) {
        self.nodeStaleAfter = nodeStaleAfter
        self.nodeExpireAfter = nodeExpireAfter
        self.edgeStaleAfter = edgeStaleAfter
        self.edgeExpireAfter = edgeExpireAfter
    }

    @discardableResult
    public func addNode(_ nodeId: NodeId, t: Double) -> NodeId {
        if let existing = nodes[nodeId.value] {
            existing.lastSeen = t
        } else {
            nodes[nodeId.value] = NodeEntry(id: nodeId, lifecycle: .discovered,
                                            lastSeen: t, firstSeen: t)
        }
        return nodeId
    }

    public func touchNode(_ nodeId: NodeId, t: Double) {
        nodes[nodeId.value]?.lastSeen = t
    }

    public func hasNode(_ nodeId: NodeId) -> Bool { nodes[nodeId.value] != nil }
    public func getNode(_ nodeId: NodeId) -> NodeEntry? { nodes[nodeId.value] }
    public func nodeIds() -> [NodeId] { nodes.values.map { $0.id } }

    @discardableResult
    public func addEdge(_ u: NodeId, _ v: NodeId, t: Double) -> (NodeId, NodeId)? {
        if nodes[u.value] == nil { addNode(u, t: t) }
        if nodes[v.value] == nil { addNode(v, t: t) }
        let key = Self.edgeKey(u, v)
        if let existing = edges[key] {
            if existing.lifecycle == .expired { existing.lifecycle = .seen }
            existing.lastSeen = t
            return existing.endpoints
        }
        let endpoints = Self.edgeEndpoints(u, v)
        edges[key] = EdgeEntry(endpoints: endpoints, lifecycle: .seen,
                               firstSeen: t, lastSeen: t)
        return endpoints
    }

    @discardableResult
    public func touchEdge(_ u: NodeId, _ v: NodeId, t: Double, bidirectional: Bool) -> Bool {
        guard let st = edges[Self.edgeKey(u, v)] else { return false }
        st.lastSeen = t
        if bidirectional { st.lastBidirectionalExchange = t }
        refreshLifecycles(t)
        return true
    }

    public func getEdge(_ u: NodeId, _ v: NodeId) -> EdgeEntry? {
        edges[Self.edgeKey(u, v)]
    }

    public func isEdgeLive(_ u: NodeId, _ v: NodeId) -> Bool {
        guard let st = getEdge(u, v) else { return false }
        return st.lifecycle != .expired && isLive(st.endpoints.0) && isLive(st.endpoints.1)
    }

    public func neighbors(_ v: NodeId) -> [NodeId] {
        var out = Set<NodeId>()
        for st in edges.values {
            guard st.lifecycle != .expired else { continue }
            let (a, b) = st.endpoints
            guard isLive(a), isLive(b) else { continue }
            if a.value == v.value { out.insert(b) }
            else if b.value == v.value { out.insert(a) }
        }
        return out.sorted()
    }

    public func isLive(_ nid: NodeId) -> Bool {
        nodes[nid.value]?.lifecycle != .expired
    }

    public func refresh(t: Double) {
        refreshLifecycles(t)
    }

    private func refreshLifecycles(_ t: Double) {
        for st in nodes.values {
            // NOT skipped when .expired: if the node has since been freshly
            // touched (authenticated radio contact resumed) its age drops and
            // it is re-admitted (relationship memory); untouched nodes stay
            // expired because their age is still past the hard threshold.
            let age = t - st.lastSeen
            let newLife: NodeLifecycle =
                age > nodeExpireAfter ? .expired : (age > nodeStaleAfter ? .stale : .active)
            if newLife != st.lifecycle {
                st.lifecycle = newLife
                if newLife == .expired {
                    for e in edges.values where e.endpoints.0 == st.id || e.endpoints.1 == st.id {
                        e.lifecycle = .expired
                    }
                }
            }
        }
        for e in edges.values {
            // EXPIRED edges stick until addEdge explicitly re-arms them (.seen)
            // on a fresh received signal (same rule as the Python engine).
            guard e.lifecycle != .expired else { continue }
            let age = t - e.lastSeen
            e.lifecycle = age > edgeExpireAfter ? .expired
                : (age > edgeStaleAfter ? .aging : .active)
        }
    }

    private static let sep: Character = "\u{0}"
    static func edgeKey(_ u: NodeId, _ v: NodeId) -> String {
        let (a, b) = u.value <= v.value ? (u.value, v.value) : (v.value, u.value)
        return "\(a)\(sep)\(b)"
    }
    static func edgeEndpoints(_ u: NodeId, _ v: NodeId) -> (NodeId, NodeId) {
        u.value <= v.value ? (u, v) : (v, u)
    }
}

File: android/app/src/main/kotlin/com/findus/crowdcompass/engine/DynamicGraph.kt
package com.findus.crowdcompass.engine

/**
 * Device-local communication graph port of core/graph.py (spec §18-20).
 * Holds only what THIS device hears: neighbors + their edges (O(k), never O(N)).
 */
class NodeEntry(val id: NodeId, var lifecycle: NodeLifecycle,
                var lastSeen: Double, val firstSeen: Double)

class EdgeEntry(val endpoints: Pair<NodeId, NodeId>, var lifecycle: EdgeLifecycle,
                val firstSeen: Double, var lastSeen: Double,
                var lastBidirectionalExchange: Double = 0.0)

class DynamicGraph(
    val nodeStaleAfter: Double = 60.0,
    val nodeExpireAfter: Double = 120.0,
    val edgeStaleAfter: Double = 60.0,
    val edgeExpireAfter: Double = 60.0,
) {
    private val nodes = linkedMapOf<String, NodeEntry>()
    private val edges = linkedMapOf<String, EdgeEntry>()

    fun addNode(nodeId: NodeId, t: Double): NodeId {
        val nid = NodeId(nodeId.value)
        val existing = nodes[nid.value]
        if (existing == null) {
            nodes[nid.value] = NodeEntry(nid, NodeLifecycle.DISCOVERED, t, t)
        } else {
            existing.lastSeen = t
        }
        return nid
    }

    fun touchNode(nodeId: NodeId, t: Double) {
        nodes[nodeId.value]?.lastSeen = t
    }

    fun hasNode(nodeId: NodeId): Boolean = nodes.containsKey(nodeId.value)

    fun getNode(nodeId: NodeId): NodeEntry? = nodes[nodeId.value]

    fun nodeIds(): List<NodeId> = nodes.values.map { it.id }

    /** Join-on-first-relationship: an unheard node is created the first time it is seen. */
    fun addEdge(u: NodeId, v: NodeId, t: Double): Pair<NodeId, NodeId>? {
        if (!nodes.containsKey(u.value)) addNode(u, t)
        if (!nodes.containsKey(v.value)) addNode(v, t)
        val key = edgeKey(u, v)
        val existing = edges[key]
        if (existing != null) {
            if (existing.lifecycle == EdgeLifecycle.EXPIRED) {
                existing.lifecycle = EdgeLifecycle.SEEN
            }
            existing.lastSeen = t
            return existing.endpoints
        }
        val endpoints = edgeEndpoints(u, v)
        edges[key] = EdgeEntry(endpoints, EdgeLifecycle.SEEN, t, t)
        return endpoints
    }

    fun touchEdge(u: NodeId, v: NodeId, t: Double, bidirectional: Boolean): Boolean {
        val key = edgeKey(u, v)
        val st = edges[key] ?: return false
        st.lastSeen = t
        if (bidirectional) st.lastBidirectionalExchange = t
        refreshLifecycles(t)
        return true
    }

    fun getEdge(u: NodeId, v: NodeId): EdgeEntry? = edges[edgeKey(u, v)]

    fun hasEdge(u: NodeId, v: NodeId): Boolean = edges.containsKey(edgeKey(u, v))

    fun isEdgeLive(u: NodeId, v: NodeId): Boolean {
        val st = getEdge(u, v) ?: return false
        return st.lifecycle != EdgeLifecycle.EXPIRED &&
            isLive(st.endpoints.first) && isLive(st.endpoints.second)
    }

    fun neighbors(v: NodeId): List<NodeId> {
        val out = linkedSetOf<String>()
        for (st in edges.values) {
            if (st.lifecycle == EdgeLifecycle.EXPIRED) continue
            val (a, b) = st.endpoints
            if (!isLive(a) || !isLive(b)) continue
            when (v.value) {
                a.value -> out.add(b.value)
                b.value -> out.add(a.value)
            }
        }
        return out.map { NodeId(it) }
    }

    fun isLive(nid: NodeId): Boolean =
        nodes[nid.value]?.lifecycle != NodeLifecycle.EXPIRED

    /** Refresh lifecycles: silence ages, signaling refreshes (spec §19-20). */
    fun refresh(t: Double) {
        refreshLifecycles(t)
    }

    private fun refreshLifecycles(t: Double) {
        for (st in nodes.values) {
            // NOT skipped when EXPIRED: freshly-touched nodes (authenticated
            // radio contact resumed) drop back under the thresholds and are
            // re-admitted (relationship memory); untouched ones stay EXPIRED
            // since their age is still past the hard threshold.
            val age = t - st.lastSeen
            val newLife = when {
                age > nodeExpireAfter -> NodeLifecycle.EXPIRED
                age > nodeStaleAfter -> NodeLifecycle.STALE
                else -> NodeLifecycle.ACTIVE
            }
            if (newLife != st.lifecycle) {
                st.lifecycle = newLife
                if (newLife == NodeLifecycle.EXPIRED) {
                    for (e in edges.values) {
                        if (e.endpoints.first.value == st.id.value ||
                            e.endpoints.second.value == st.id.value
                        ) {
                            e.lifecycle = EdgeLifecycle.EXPIRED
                        }
                    }
                }
            }
        }
        for (e in edges.values) {
            // EXPIRED edges stick until addEdge explicitly re-arms them (.SEEN)
            // on a fresh received signal (same rule as the Python engine).
            if (e.lifecycle == EdgeLifecycle.EXPIRED) continue
            val age = t - e.lastSeen
            val newLife = when {
                age > edgeExpireAfter -> EdgeLifecycle.EXPIRED
                age > edgeStaleAfter -> EdgeLifecycle.AGING
                else -> EdgeLifecycle.ACTIVE
            }
            if (newLife != e.lifecycle) e.lifecycle = newLife
        }
    }

    companion object {
        private const val SEP = '\u0000'
        fun edgeKey(u: NodeId, v: NodeId): String {
            val (a, b) = if (u.value <= v.value) u.value to v.value else v.value to u.value
            return a + SEP + b
        }
        fun edgeEndpoints(u: NodeId, v: NodeId): Pair<NodeId, NodeId> =
            if (u.value <= v.value) u to v else v to u
    }
}